20260106_095917_Billes_170nm_YellowGreen/GUI_parameters.txt
{
    "cameras": {
        "camera_0": {
            "camera_id": 0,
            "hchipsize": 4432,
            "vchipsize": 2368,
            "pixel_size": 4.6000000000000005,
            "sample_pixel_size": 0.15535292131036815,
            "hsize": 2048,
            "hpos": 1192,
            "vsize": 2048,
            "vpos": 160,
            "exposure_time": 0.02,
            "binning": 1,
            "line_readout_time": 7.3090909090909095e-06,
            "image_readout_time": 0.007542981818181819
        }
    },
    "channels": {
        "channel_GFP": {
            "channel_id": "GFP",
            "is_active": true,
            "channel_order": null,
            "laser_is_active": {
                "405": false,
                "488": true,
                "561": false,
                "640": false
            },
            "laser_power": {
                "405": 0,
                "488": 20,
                "561": 0,
                "640": 0
            },
            "filter": "GFP",
            "camera": 0,
            "exposure_time": 20.0
        }
    },
    "experiment": {
        "exp_name": "Billes_170nm_YellowGreen",
        "data_path": "",
        "timepoints": 1,
        "time_intervals": 1.0,
        "total_duration": 1.0,
        "mode": "standard",
        "scanner_position": 0,
        "scan_range": 299.9394782925777,
        "aspect_ratio": 0.8390996311772799,
        "n_steps": 1480,
        "step_size": 0.20279883589761846,
        "slit_aperture": 800
    },
    "microscope": {
        "tilt_angle": 40.0,
        "mag_total": 29.61,
        "volts_per_um": 0.01024,
        "galvo_response_time": 0.3,
        "galvo_flyback_time": 3.0,
        "stage_port": "COM10",
        "filter_changing_time": [
            55,
            70,
            85
        ],
        "filters": [
            "BFP",
            "GFP",
            "CY3.5",
            "TexRed",
            "empty5",
            "empty6"
        ],
        "lasers": [
            "405",
            "488",
            "561",
            "640"
        ],
        "volts_per_laser_percent": {
            "405": 0.05,
            "488": 0.05,
            "561": 0.05,
            "640": 0.05
        },
        "laser_response_time": 0.02,
        "OxxiusCombiner_port": "COM5",
        "OxxiusCombiner_command": {
            "405": null,
            "488": null,
            "561": "PPL2",
            "640": null
        },
        "daq_channels": {
            "co_channel": "Dev1/ctr0",
            "co_terminal": "/Dev1/PFI0",
            "galvo": "Dev1/ao0",
            "camera_0": "Dev1/port0/line0",
            "camera_1": "Dev1/port0/line1",
            "filter_wheel_1": "Dev1/port0/line2",
            "filter_wheel_2": "Dev1/port0/line3",
            "channel_finished": "/Dev1/PFI1",
            "laser_blanking": "Dev1/port0/line4"
        },
        "daq_channels_laser_analog_out": {
            "405": "Dev1/ao1",
            "488": "Dev1/ao2",
            "561": null,
            "640": "Dev1/ao3"
        },
        "daq_channels_laser_digital_out": {
            "405": "Dev1/port0/Line8",
            "488": "Dev1/port0/Line9",
            "561": "Dev1/port0/Line10",
            "640": "Dev1/port0/Line11"
        }
    }
}

configs/config.py
# -*- coding: utf-8 -*-
"""
Created on Fri Feb 14 09:26:31 2025

@author: tbrugiere
"""

LASERS = ["405","488","561","640"]

class camera(object):
    "Object discribing camera settings"
    # TODO: get camera brand, model, serial number 
    def __init__(self, camera_id):
        self.camera_id = camera_id
        self.hchipsize = 4432
        self.vchipsize = 2368
        self.pixel_size = 4.6 #in µm
        self.sample_pixel_size = 0.153 # in µm, calculated later
        self.hsize = 4432
        self.hpos = 0
        self.vsize = 2308
        self.vpos = 0
        self.binning = 1
        self.exposure_time = 0.01 # in seconds
        self.line_readout_time = 7.309E-06 # (s) temps de lecture/ligne (7.309E-06 lignes d'aprés la doc)

        self.calculate_image_readout_time()
        
    def calculate_image_readout_time(self):            
        self.image_readout_time = (self.vsize / 2 + 8) * self.line_readout_time

class channel_config(object):
    "Object discribing channels settings"
    def __init__(self, channel_id, lasers = LASERS): #lasers = ["405","488","561","640"]
        self.channel_id = channel_id # channel_id : nom du canal de type chr
        self.is_active = False
        self.channel_order = None

        # Channel set from the interface
        self.laser_is_active = {}
        self.laser_power = {}
        self.filter = 'BFP' #Name of the filter in the weel
        self.camera = 0 #Camera ID
        self.exposure_time = 10 #Exposure time in ms
        
        for laser in lasers:
            self.laser_is_active[laser] = False
            self.laser_power[laser] = 0 # laser_power en %
            

class experiment(object):
    "Object discribing experiment settings"
    def __init__(self):
        self.exp_name = 'Image'
        self.data_path = "D:/Projets_Python/OPM_GUI"
        self.timepoints = 10
        self.time_intervals = 1.0 # in s
        self.total_duration = 10.0 # in s
        self.mode = "standard"
        
        self.scanner_position = 0
        self.scan_range = 20
        self.aspect_ratio = 3
        self.n_steps = 0 # Calculé plus tard
        self.step_size = 0 # Calculé plus tard
        
        self.slit_aperture = 800 #in µm
        
        self.cameras = []
        self.channels = []
        
class microscope(object):
    def __init__(self):
        
        self.tilt_angle = 40.0 #in degrees
        self.mag_total = 29.61
        
        # scanning galvo
        self.volts_per_um = 0.01024
        self.galvo_response_time = 0.3  # A mesure in ms
        self.galvo_flyback_time = 3.0 # A mesurer in ms
        
        # stage
        self.stage_port = 'COM10'
        
        # Filters
        self.filter_port = "COM?"
        self.filter_changing_time = [55,70,85] # in ms
        self.filters = ['BFP','GFP','CY3.5','TexRed','empty5', 'empty6'] #should be 6 options
        
        # Lasers
        self.lasers = LASERS # ["405","488","561","640"]
        self.volts_per_laser_percent = {'405' : 5.0 / 100.0, # 5v max
                                        '488' : 5.0 / 100.0, # 5v max
                                        '561' : 5.0 / 100.0, # 5v max
                                        '640' : 5.0 / 100.0  # 5v max
                                        }
        self.laser_response_time = 20e-3 # in ms
        self.OxxiusCombiner_port = "COM5"
        
        self.OxxiusCombiner_command = {"405" : None, # Régle digital modulation (dm) du laser 405
                                       "488" : None, # Régle digital modulation (dm) du laser 488
                                       "561" : "PPL2", # Régle digital modulation (dm) du laser 561
                                       "640" : None, # Régle digital modulation (dm) du laser 640
                                       } #Command send to the combiner to set laser power,

        # DAQ
        self.daq_channels = {"co_channel": "Dev1/ctr0", # ADD: trigger start of each volume
                             "co_terminal": "/Dev1/PFI0", # ADD: trigger start of each volume
                             "galvo": "Dev1/ao0", # Fait bouger le galvo pour le scanning
                             "camera_0": "Dev1/port0/line0", # Trigger l'exposition de la camera
                             "camera_1": "Dev1/port0/line1", # Trigger l'exposition de la deuxieme camera (si presente)
                             "filter_wheel_1": "Dev1/port0/line2", # Trigger de la roue de filtres
                             "filter_wheel_2": "Dev1/port0/line3", # Trigger de la roue de filtres
                             "channel_finished" : "/Dev1/PFI1", # Trigger at the end of each channel for the computer (not used yet)
                             "laser_blanking" : "Dev1/port0/line4", # Trigger le blanking de l'AOTF du banc laser
                             }
        
        self.daq_channels_laser_analog_out = {"405" : "Dev1/ao1", # Régle la puissance du laser 405
                                              "488" : "Dev1/ao2",      # Régle la puissance du laser 488
                                              "561" : None, # Régle la puissance du laser 561
                                              "640" : "Dev1/ao3", # Régle la puissance du laser 640
                                              }
        
        self.daq_channels_laser_digital_out = {"405" : "Dev1/port0/Line8", # Régle digital modulation (dm) du laser 405
                                               "488" : "Dev1/port0/Line9", # Régle digital modulation (dm) du laser 488
                                               "561" : "Dev1/port0/Line10", # Régle digital modulation (dm) du laser 561
                                               "640" : "Dev1/port0/Line11", # Régle digital modulation (dm) du laser 640
                                               }
        
        ##############################
        ## Microscope configuration ##
        ##############################
        
        # Objectives
        
            ## primary objective
        self.OBJ_1_MANUFACTURER = "Nikon"
        self.OBJ_1_MODEL = "CFI75 Apo 25XC W"
        self.OBJ_1_SN = "0000000"
        self.OBJ_1_MAGNIFICATION = 25
        self.OBJ_1_NA = 1.1
        self.OBJ_1_EFL = 8  # in mm
        self.OBJ_1_IM = "Water dipping"
        self.OBJ_1_IM_RI = 1.333
        
            ## secondary objective
        self.OBJ_1_MANUFACTURER = "Nikon"
        self.OBJ_1_MODEL = "CFI Plan Apochromat Lambda D 20X"
        self.OBJ_1_SN = "0000000"
        self.OBJ_1_MAGNIFICATION = 20
        self.OBJ_1_NA = 0.8
        self.OBJ_1_EFL = 10  # in mm
        self.OBJ_1_IM = "Air"
        self.OBJ_1_IM_RI = 1
        
            ## tertiary objective
        self.OBJ_3_MANUFACTURER = "Special Optics"
        self.OBJ_3_MODEL = "AMS-AGY v2.0"
        self.OBJ_3_SN = "0000"
        self.OBJ_3_MAGNIFICATION = 200/9
        self.OBJ_3_NA = 1.0
        self.OBJ_3_EFL = 9  # in mm
        self.OBJ_3_IM = "Glass"
        self.OBJ_3_IM_RI = 1.525
        #self.OBJ3_PIEZO_POS_UM = 0 #Dans le fichier original
        
        # Lenses
        
            ## tube lens 1
        self.TUBE_LENS_1_MANUFACTURER = "Thorlabs"
        self.TUBE_LENS_1_MODEL = "TTL200MP"
        self.TUBE_LENS_1_EFL = 200  # in mm

            ## tube lens 2
        self.TUBE_LENS_2_MANUFACTURER = "Thorlabs"
        self.TUBE_LENS_2_MODEL = "EFL188"
        self.TUBE_LENS_2_EFL = 188  # in mm

            ## tube lens 3
        self.TUBE_LENS_3_MANUFACTURER = "Nikon"
        self.TUBE_LENS_3_MODEL = "Nikon 200mm Tube Lens "
        self.TUBE_LENS_3_EFL = 200  # in mm

            ## scan lens 1
        self.SCAN_LENS_1_MANUFACTURER = "Thorlabs"
        self.SCAN_LENS_1_MODEL = "CLS-SL"
        self.SCAN_LENS_1_EFL = 70

            ## scan lens 2
        self.SCAN_LENS_2_MANUFACTURER = "Thorlabs"
        self.SCAN_LENS_2_MODEL = "CLS-SL"
        self.SCAN_LENS_2_EFL = 70
        
        # Filters, dichroics
            ## Filter weel
            
            ## Filters
            
            ## Dichroic
            
    def to_dict(self):
        
        microscope_dict = {
            "tilt_angle" : self.tilt_angle,
            "mag_total" : self.mag_total,
            # scanning galvo
            "volts_per_um" : self.volts_per_um,
            "galvo_response_time" : self.galvo_response_time,  # A mesure in ms
            "galvo_flyback_time" : self.galvo_flyback_time, # A mesurer in ms
            
            # stage
            "stage_port" : self.stage_port,
            
            # Filters
            "filter_port" : self.filter_port,
            "filter_changing_time" : self.filter_changing_time, # in ms
            "filters" : self.filters, #should be 6 options
            
            # Lasers
            "lasers" : self.lasers, # ["405","488","561","640"]
            "volts_per_laser_percent" : self.volts_per_laser_percent, # 5v max
            "laser_response_time" : self.laser_response_time, # in ms
            "OxxiusCombiner_port" : self.OxxiusCombiner_port,
            "OxxiusCombiner_command" : self.OxxiusCombiner_command,
            
            # NiDAQ
            "daq_channels" : self.daq_channels,
            "daq_channels_laser_analog_out" : self.daq_channels_laser_analog_out,
            "daq_channels_laser_digital_out" : self.daq_channels_laser_digital_out
            }
        
        return microscope_dict
        
    def from_dict(self, microscope_dict):

        self.tilt_angle = microscope_dict["tilt_angle"]
        self.mag_total = microscope_dict["mag_total"]
        self.volts_per_um = microscope_dict["volts_per_um"]
        self.galvo_response_time = microscope_dict["galvo_response_time"]
        self.galvo_flyback_time = microscope_dict["galvo_flyback_time"]
        self.stage_port = microscope_dict["stage_port"]
        self.filter_port = microscope_dict["filter_port"]
        self.filter_changing_time = microscope_dict["filter_changing_time"]
        self.filters = microscope_dict["filters"]
        self.lasers = microscope_dict["lasers"]
        self.volts_per_laser_percent = microscope_dict["volts_per_laser_percent"]
        self.laser_response_time = microscope_dict["laser_response_time"]
        self.OxxiusCombiner_port = microscope_dict["OxxiusCombiner_port"]
        self.OxxiusCombiner_command = microscope_dict["OxxiusCombiner_command"]
        self.daq_channels = microscope_dict["daq_channels"]
        self.daq_channels_laser_analog_out = microscope_dict["daq_channels_laser_analog_out"]
        self.daq_channels_laser_digital_out = microscope_dict["daq_channels_laser_digital_out"]